### ISSUE TITLE
Maximum recursion depth when computing the maximum independent set

### ISSUE THREAD

**Issue by huku-** [networkx/networkx#3455]
Hello,

I get a `RuntimeError` (maximum recursion depth exceeded) when using `maximum_independent_set()` on graphs with self-loops. Here's a small test case for reproducing this issue:

```python
from networkx.algorithms import approximation

g = networkx.DiGraph()
g.add_node(1)
g.add_node(2)
g.add_node(3)
g.add_edge(1, 2)
g.add_edge(2, 2)
g.add_edge(2, 3)

print approximation.independent_set.maximum_independent_set(g)
```

The problem seems to be in `ramsey_R2()`, but don't take my word for it as I haven't fully debugged the problem yet :-) Looks like the self-loop node always ends up in `nbrs` and, consequently, the recursion's termination condition is never reached.

**Comment by dschult** [comment]
It's quite possible that the Ramsey_R2 code didn't take selfloops into account when it was written. As such, I think we can call this a bug. Do you see a good way to either identify the problem and raise an error, or even better, fix the error?

**Comment by huku-** [comment]
Changing the following line in **ramsey.py**:

```python
nbrs = nx.all_neighbors(G, node)
```

To something like the following:

```python
nbrs = set(nx.all_neighbors(G, node))
nbrs.discard(node)
```

Seems to be doing the trick. I know it's not very elegant, as it drops the benefits of the iterator returned by `all_neighbors()`, but has worked fine for me so far.

**Comment by dschult** [comment]
Great!  Thanks --

I have a suggestion to keep it as an iterator:

    nbrs = (nbr for nbr in nx.all_neighbors(G, node) if nbr != node)

And we should add a line of doc_string to explain that we ignore self-loops.
Thanks again!

**Comment by dschult** [comment]
Fixed in #4037
